src/Core/DevOps/StaticAnalyze/PHPStan/Rules/NoUselessVarAnnotationRule.php

```
@@ -0,0 +1,132 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\DevOps\StaticAnalyze\PHPStan\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Stmt\Expression;
+use PHPStan\Analyser\NameScope;
+use PHPStan\Analyser\Scope;
+use PHPStan\PhpDoc\TypeNodeResolver;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocNode;
+use PHPStan\PhpDocParser\Lexer\Lexer;
+use PHPStan\PhpDocParser\Parser\PhpDocParser;
+use PHPStan\PhpDocParser\Parser\TokenIterator;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\Type;
+use PHPStan\Type\VerbosityLevel;
+use Shopware\Core\Framework\Log\Package;
+
+/**
+ * @internal
+ *
+ * @implements Rule<Expression>
+ */
+#[Package('framework')]
+class NoUselessVarAnnotationRule implements Rule
```

I wonder if we cannot use already existing rules of PHPStan instead of introducing something new :thinking: 

this seems to match your use case https://phpstan.org/config-reference#reportwrongphpdoctypeinvartag

by default PHPStan only checks native types, so maybe we can simply remove https://github.com/shopware/shopware/blob/trunk/src/Core/DevOps/StaticAnalyze/PHPStan/common.neon#L21 to achieve the same

---

also the next config sounds interesting https://phpstan.org/config-reference#reportanytypewideninginvartag

---

In the case of the new phpstan rule i intrduced with this PR the rule checks for useless casts. this means phpstan comments which use the same type as already nativly provided. e.g.:

$s = 'string';
/** @var string */
$s2 = $s;

would be catched by this rule. but this doesn't get catched by the other phpstan rules, as they only check for wrong types (in this case the type is not wrong, it is just redundant)

---

@mitelg I enabled `reportAnyTypeWideningInVarTag` and resolved the issues

`reportWrongPhpDocTypeInVarTag` is too much, there are almost 900 errors when this rule is enabled.

---

okay, I understand :+1: 

but still I do not see this in Shopware. this should be more a general part of PHPStan itself. I think this is something that could be beneficial for every user of PHPStan

---

I'll remove the custom rule with the next commit

---

maybe we can enable `reportWrongPhpDocTypeInVarTag` and ignore the errors with different paths like you did with the `EntityRepository` stuff :thinking:

---

with that, we can tackle them one by one

---

@mitelg We could do this, but this would be for another PR